src/vision/status_quo/alan_tries_a_socket_sink.md

```diff
@@ -0,0 +1,106 @@
+# 😱 Status quo stories: Alan tries using a socket Sink
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+Alan is working on a project that uses an alternative popular async framework. He finds the new async framework less familiar, and as he reads through their GitHub issues, he's saddened by the project maintainer's unwelcoming attitude towards their community. He asks his teammates if they can switch to the one he's more familiar with and feels more welcome in, but the teammates are understandably resistant to changing something that important just for one feature. So, he works hard to find workarounds to accomplish his team's goals.
+
+One of the goals is to switch from a WebSocket implementation using raw TCP sockets to one managed behind an HTTP server library, so both HTTP and WebSocket commands can be forwarded to a transport-agnostic RPC server. He finds an HTTP server that's similar to one he's used to using with the other async framework, and a WebSocket middleware library that goes with it.
+
+However, as he's working, Alan encounters a situation where the socket needs to be written to within an async thread, and the traits just aren't working. He wants to split the stream into a sender and receiver:
```

Comment:
I'm having a hard time understanding this, actually, there seems to be some context I am missing. Poking at the rustdoc, I see the [`split`](https://docs.rs/futures/0.3.13/futures/stream/trait.StreamExt.html#method.split) method, that works on a `Stream + Sink`. Is the reason that this code doesn't work because `ws_stream` does not implement `Sink`? That could be made clearer -- for example, you could show the error message he gets.

Reply:
It's not really an error message so much as undesired locking behavior.

Reply:
OK, I think maybe I'm starting to see. From reading the code snippet, what I get is this:

* He wants to have incoming messages and then spawn off threads that will do some RPC before forwarding the messages on to another place?

I think what's confusing to me is that I think of a stream as an "async iterator", and I don't quite get how it could be split. It seems like what you want is to "splice" yourself in the middle of the stream?

Reply:
Yep! I think `async iterator` is a good way to think of a stream as long as it's coupled with the perspective that new items can be added to the stream, I don't think it has to be finite. It's kinda like a duplex unbounded MPSC. But streams vary a great deal on their implementation details, such as high water mark, and backpressure signalling, but that's a little beyond the scope of this.

That said, I'm coming at this from a NodeJS perspective, where stream use is common, and often preferred. A WebSocket is an ideal candidate for this, but it's important it can be both read and written to in exclusion.

Reply:
Iterators don't have to be finite either, right?

Ultimately I think of iterators (and streams) as "producers" and functions (and sinks) as "consumers".  I was finding it very odd to think about something being `Stream + Sink`, but I guess it makes sense if you think of (say) a network port, or a bidirectional channel. I guess we are talking about a *socket* here, so that makes sense.

Reply:
Yep, I suppose you're right!

Reply:
I'd love some clarification on the splice part of this. If there were a way to split the connection into a "sender" and a "receiver" wherein the receiver is Stream and the sender just has a send(&self, message: Message) associated function but which doesn't implement Sink, what would be missing?

As a separate question, if the same type that was the Stream also had the send function and was Clone so any clone of it could act as either side of that, what would be missing?

Reply:
So, if it's cloned, it wouldn't have to be locked by the sender thread, while also being read by the receiver loop?

Reply:
Also, I'm totally for improvements to `tide-websockets`, but is that not outside the scope of this PR?

Reply:
It's worth figuring out if the Sink trait is necessary, since that's what this issue and the closed issue were about.

The latter situation I described where any clone could act as sender or receiver wouldn't require an improvement, it's the current state of the code. The sender and receiver are locked internally, so there can only be one clone sending at a time, and one clone polling the stream at a time, but they can be sending and receiving concomitantly

Reply:
I think the splicing was a misunderstanding on my part. I eventually realized that what was happening was just receiving messages, doing some processing, and responding in kind.

Reply:
Definitely changes to tide-websockets or whatever are not in scope for the story.

Reply:
I don't think the *story itself* should describe whether this can be solved without `Sink`, but I would definitely love to explore possible resolutions to this story as "shiny futures", or possibly as FAQs. I think we can merge without that. I talked to @cryptoquick some -- one thing that I was explaining is that it should be possible for him to add a newtype wrapped around the `WebSocket` type that implements `futures::Sink` and `futures::Stream`, which would then allow him to write his original example without having to maintain a complete parallel fork of the crate. This seems better to me.

It does seem to me that *some* trait that means "can consume values of type `Item`" would be appropriate. I am not convinced that the `Sink` trait as is the correct design for this.